src/package-operations.ts

```diff
@@ -313,41 +321,42 @@ function getUpdatedDependencyField(
  * @param fieldsToValidate - The manifest fields that will be validated.
  * @returns The object corresponding to the parsed package.json file.
  */
-export async function getPackageManifest<T extends keyof PackageManifest>(
+export async function getPackageManifest<T extends FieldNames>(
   containingDirPath: string,
-  fieldsToValidate?: T[],
-): Promise<Pick<PackageManifest, T>> {
-  const manifest = await readJsonFile(
+  fieldsToValidate: T[],
+): Promise<Pick<PackageManifest, T> & Partial<PackageManifest>> {
+  const manifest: Partial<PackageManifest> = await readJsonObjectFile(
     pathUtils.join(containingDirPath, PACKAGE_JSON),
   );
 
-  validatePackageManifest(manifest, containingDirPath, fieldsToValidate);
-  return manifest as Pick<PackageManifest, T>;
+  return validatePackageManifest(containingDirPath, manifest, fieldsToValidate);
 }
 
 /**
- * Validates a manifest by ensuring that the given fields are properly formatted
- * if present. Fields that are required by the `PackageManifest` interface must be
- * present if specified.
+ * Validates a manifest by ensuring that the given fields are present and
+ * properly formatted.
  *
  * @see PackageManifest - For fields that must be present if specified.
- * @param manifest - The manifest to validate.
  * @param manifestDirPath - The path to the directory containing the
- * package.json file.
+ * manifest file.
+ * @param manifest - The manifest to validate.
  * @param fieldsToValidate - The manifest fields that will be validated.
+ * @returns The unmodified manifest, with validated fields typed correctly.
  */
-export function validatePackageManifest(
-  manifest: Partial<PackageManifest>,
+export function validatePackageManifest<
+  T extends Partial<PackageManifest>,
```

Comment:
I remain a bit confused about why these generic parameters are necessary. It seems that there are two types of manifests that we want to validate: a monorepo manifest, and a non-monorepo manifest. Why do we need _two_ generic parameters, and to pass in which fields to validate? This all seems a bit... overly complicated? Maybe I'm missing something

Reply:
It was complicated by the fact that there's effectively _three_ different types of package manifests, going from subset to superset.

The first is the root manifest, which could be either a normal or a monorepo `package.json` file. There, we only need to validate the `version`. For the polyrepo case, we also need to validate `name`. For the monorepo case, we need to also validate `private` and `workspaces`. On top of that, deeper in the monorepo workflow, we need to validate `version` and `name` for the manifests of the monorepo's packages.

Rather than keeping a single generic function around, I split the validation logic out of `getPackageManifest` and created separate validation function for all the different cases. Most of these new functions use a single generic, but it's a convenient to take in e.g. a `Partial` manifest object and return it with e.g. one field correctly type.

I think this is probably better, because the generic validation function kind of sucked.

Reply:
Part of what I was missing is that you were aiming to perform the absolute minimum validation necessary. I would have probably used just a single validation function that validated `name` and `version`, and also conditionally validated `workspaces` and `private` if they existed (and that `private` was `true` if `workspaces` was present). That'd give us one `PackageManifest` type. Then a simple check that `workspaces` existed would give us a second type, `MonorepoPackageManifest`.

But skipping the name validation for the root does make that a bit more challenging.

Reply:
> But skipping the name validation for the root does make that a bit more challenging.

Yeah, validating the `name` would require that we mandate that monorepo root manifests have a placeholder name field, which seems kind of bad.